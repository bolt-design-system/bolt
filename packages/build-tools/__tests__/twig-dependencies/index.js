const { render } = require('@bolt/twig-renderer');
const path = require('path');
const fs = require('fs-extra');
const shell = require('shelljs');

let generatedTwigNamespaces;
let generatedCSS;

const {
  startServer,
  stopServer,
} = require('@bolt/testing-server/server-only.js');

describe('Build Deps Resolution', () => {
  let page;

  afterAll(async () => {
    await stopServer();
    await page.close();
  });

  beforeEach(async () => {
    await page.evaluate(() => {
      document.body.innerHTML = '';
    });
  });

  beforeAll(async () => {
    await shell.exec(`bolt build --prod --config-file .boltrc.js`, {
      silent: true,
    });

    const serverPath = await startServer({
      wwwDir: './dist',
    });

    page = await global.__BROWSER__.newPage();
    await page.goto(serverPath, {
      timeout: 0,
    });

    generatedTwigNamespaces = JSON.parse(
      fs.readFileSync(
        path.resolve(
          process.cwd(),
          `./dist/build/data/twig-namespaces.bolt.json`,
        ),
      ),
    );

    generatedCSS = fs.readFileSync(
      path.resolve(process.cwd(), `./dist/build/bolt-global.css`),
    );
  });

  test('compiled bolt-global.css includes styles for other global Bolt components listed in package.json', async () => {
    expect(generatedCSS.includes('.c-bolt-ratio')).toBe(true);
  });

  test('compiled CSS does NOT contain styles for non-global Bolt components', async () => {
    expect(generatedCSS.includes('bolt-carousel')).toBe(false);
  });

  test('auto-generated Twig namespaces includes Bolt components listed in the package.json', async () => {
    expect(
      Object.keys(generatedTwigNamespaces).includes('bolt-components-ratio'),
    ).toBe(true);
  });

  test('auto-generated Twig namespaces includes individual Bolt components', async () => {
    expect(
      Object.keys(generatedTwigNamespaces).includes('bolt-components-carousel'),
    ).toBe(true);
  });

  test('compiled code with auto-injected dependencies renders correctly', async () => {
    const results = await render('@bolt-components-image/image.twig', {
      src: '/fixtures/1200x660.jpg',
      alt: 'A Rock Climber',
      lazyload: true,
    });

    const htmlResult = results.html;

    await page.evaluate(async htmlResult => {
      document.body.insertAdjacentHTML('beforeend', `${htmlResult}`);
    }, htmlResult);

    await page.waitFor(500);

    const renderedHTML = await page.evaluate(async () => {
      const image = document.querySelector('bolt-image');
      await customElements.whenDefined('bolt-image');
      await image.updateComplete;
      return image.renderRoot.innerHTML;
    });

    expect(results.ok).toBe(true);
    expect(results.html).toMatchSnapshot();
    expect(renderedHTML).toMatchSnapshot();

    const image = await page.screenshot();
    expect(image).toMatchImageSnapshot();
  });
});
