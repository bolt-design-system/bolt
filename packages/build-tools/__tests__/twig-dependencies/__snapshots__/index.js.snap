// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Build Deps Resolution compiled code with auto-injected dependencies renders correctly 1`] = `
<bolt-image src="/fixtures/1200x660.jpg"
            srcset="/fixtures/1200x660-50.jpg 50w, /fixtures/1200x660-100.jpg 100w, /fixtures/1200x660-200.jpg 200w, /fixtures/1200x660-320.jpg 320w, /fixtures/1200x660-480.jpg 480w, /fixtures/1200x660-640.jpg 640w, /fixtures/1200x660-800.jpg 800w, /fixtures/1200x660-1024.jpg 1024w, /fixtures/1200x660.jpg 1200w"
            alt="A Rock Climber"
            sizes="auto"
            ratio="1200/660"
            placeholder-color="hsl(233, 33%, 97%)"
            placeholder-image="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=="
            style="background-color: hsl(233, 33%, 97%);"
>
  <bolt-ratio ratio="20/11">
    <replace-with-children class="c-bolt-ratio"
                           style="padding-bottom: calc((11 / 20) * 100%); --aspect-ratio: 1.8181818181818;"
    >
      <img class="c-bolt-image__image-placeholder"
           alt="A Rock Climber"
           src="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=="
      >
      <img class="c-bolt-image__image c-bolt-image__lazyload c-bolt-image__lazyload--fade c-bolt-image__lazyload--blur js-lazyload"
           alt="A Rock Climber"
           data-srcset="/fixtures/1200x660-50.jpg 50w, /fixtures/1200x660-100.jpg 100w, /fixtures/1200x660-200.jpg 200w, /fixtures/1200x660-320.jpg 320w, /fixtures/1200x660-480.jpg 480w, /fixtures/1200x660-640.jpg 640w, /fixtures/1200x660-800.jpg 800w, /fixtures/1200x660-1024.jpg 1024w, /fixtures/1200x660.jpg 1200w"
           data-sizes="auto"
           srcset="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=="
           src="/fixtures/1200x660.jpg"
      >
      <noscript>
        <img class="c-bolt-image__image"
             alt="A Rock Climber"
             src="/fixtures/1200x660.jpg"
             srcset="/fixtures/1200x660-50.jpg 50w, /fixtures/1200x660-100.jpg 100w, /fixtures/1200x660-200.jpg 200w, /fixtures/1200x660-320.jpg 320w, /fixtures/1200x660-480.jpg 480w, /fixtures/1200x660-640.jpg 640w, /fixtures/1200x660-800.jpg 800w, /fixtures/1200x660-1024.jpg 1024w, /fixtures/1200x660.jpg 1200w"
        >
      </noscript>
    </replace-with-children>
  </bolt-ratio>
</bolt-image>
`;

exports[`Build Deps Resolution compiled code with auto-injected dependencies renders correctly 2`] = `
<bolt-ratio ratio="1200/660">
  <img class="c-bolt-image__image-placeholder"
       src="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=="
       alt="A Rock Climber"
  >
  <img class="c-bolt-image__image c-bolt-image__lazyload c-bolt-image__lazyload--fade c-bolt-image__lazyload--blur lazyautosizes ls-is-cached is-lazyloaded"
       src="/fixtures/1200x660.jpg"
       alt="A Rock Climber"
       srcset="/fixtures/1200x660-50.jpg 50w, /fixtures/1200x660-100.jpg 100w, /fixtures/1200x660-200.jpg 200w, /fixtures/1200x660-320.jpg 320w, /fixtures/1200x660-480.jpg 480w, /fixtures/1200x660-640.jpg 640w, /fixtures/1200x660-800.jpg 800w, /fixtures/1200x660-1024.jpg 1024w, /fixtures/1200x660.jpg 1200w"
       data-srcset="/fixtures/1200x660-50.jpg 50w, /fixtures/1200x660-100.jpg 100w, /fixtures/1200x660-200.jpg 200w, /fixtures/1200x660-320.jpg 320w, /fixtures/1200x660-480.jpg 480w, /fixtures/1200x660-640.jpg 640w, /fixtures/1200x660-800.jpg 800w, /fixtures/1200x660-1024.jpg 1024w, /fixtures/1200x660.jpg 1200w"
       sizes="784px"
       data-sizes="auto"
       style="object-position: center center;"
  >
</bolt-ratio>
`;
